capa/main.py

```diff
@@ -224,7 +231,12 @@ def find_code_capabilities(
 def find_file_capabilities(ruleset: RuleSet, extractor: FeatureExtractor, function_features: FeatureSet):
     file_features = collections.defaultdict(set)  # type: FeatureSet
 
-    for feature, va in itertools.chain(extractor.extract_file_features(), extractor.extract_global_features()):
+    if isinstance(extractor, StaticFeatureExtractor):
+        extractor_: StaticFeatureExtractor = cast(StaticFeatureExtractor, extractor)
```

Comment:
can you explain what the `cast` is doing here? i sort of get it but im not totally sure, so an explanation would help me.

Reply:
many of such typecasts were added just to avoid mypy errors like this:

![image](https://github.com/mandiant/capa/assets/16624109/022bbedf-a2a4-4d67-9631-4b0e0f4b5821)

my belief is that I would need to typecast back from a `Union` type into one of those two classed in order to call its methods, is there a better way to do this?

Reply:
i think that `DynamicFeatureExtractor` no longer requires those methods in its interface (at least as written above) so mypy is correctly complaining here. using the cast hides the error but doesn't address it. we should probably add `extract_file_features` and `extract_global_features` to the dynamic feature interface *or* introduce different logic here to not invoke those routines on the dynamic extractor instances.

in general, i've found that using `assert isinstance(foo, bar)` is enough to hint to mypy of the expected type, and if it still complains, then its *usually* my fault.